app/controllers/declaration/SupervisingCustomsOfficeController.scala

```diff
@@ -0,0 +1,79 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.declaration
+
+import controllers.actions.{AuthAction, JourneyAction}
+import controllers.navigation.Navigator
+import forms.declaration.WarehouseDetails
+import javax.inject.Inject
+import models.requests.JourneyRequest
+import models.{ExportsDeclaration, Mode}
+import play.api.data.Form
+import play.api.i18n.I18nSupport
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import services.cache.ExportsCacheService
+import uk.gov.hmrc.play.bootstrap.controller.FrontendController
+import views.html.declaration.supervising_customs_office
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class SupervisingCustomsOfficeController @Inject()(
+  authenticate: AuthAction,
+  journeyType: JourneyAction,
+  navigator: Navigator,
+  override val exportsCacheService: ExportsCacheService,
+  mcc: MessagesControllerComponents,
+  supervisingCustomsOfficePage: supervising_customs_office
+)(implicit ec: ExecutionContext)
+    extends FrontendController(mcc) with I18nSupport with ModelCacheable {
+
+  import forms.declaration.WarehouseDetails._
+
+  def displayPage(mode: Mode): Action[AnyContent] = (authenticate andThen journeyType) { implicit request =>
+    request.cacheModel.locations.warehouseIdentification match {
+      case Some(data) => Ok(supervisingCustomsOfficePage(mode, form().fill(data)))
+      case _          => Ok(supervisingCustomsOfficePage(mode, form()))
+    }
+  }
+
+  def submit(mode: Mode): Action[AnyContent] = (authenticate andThen journeyType).async { implicit request =>
+    form()
+      .bindFromRequest()
+      .fold(
+        (formWithErrors: Form[WarehouseDetails]) => Future.successful(BadRequest(supervisingCustomsOfficePage(mode, formWithErrors))),
+        form => {
+          updateCache(form)
+            .map(_ => navigator.continueTo(controllers.declaration.routes.WarehouseDetailsController.displayPage(mode)))
+        }
+      )
+  }
+
+  private def updateCache(formData: WarehouseDetails)(implicit request: JourneyRequest[AnyContent]): Future[Option[ExportsDeclaration]] =
+    updateExportsDeclarationSyncDirect { model =>
+      val warehouseDetails = model.locations.warehouseIdentification
+        .map(
+          dbWarehouseDetails =>
+            WarehouseDetails(
+              identificationNumber = dbWarehouseDetails.identificationNumber,
+              supervisingCustomsOffice = formData.supervisingCustomsOffice,
+              inlandModeOfTransportCode = dbWarehouseDetails.inlandModeOfTransportCode
+          )
+        )
+        .getOrElse(WarehouseDetails(supervisingCustomsOffice = formData.supervisingCustomsOffice))
+      model.copy(locations = model.locations.copy(warehouseIdentification = Some(warehouseDetails)))
```

Comment:
Can we extract logic about copying to exports declaration model?

Reply:
I've spoken to Paulo about this and the model is going to be reviewed in another ticket, at the moment we are using the same model for three pages.